File: go/progress.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// Progress represents user progress
type Progress struct {
	ChapterID   string `json:"chapter_id"`
	ExerciseID  string `json:"exercise_id"`
	Completed   bool   `json:"completed"`
	Attempts    int    `json:"attempts"`
	HintsUsed   int    `json:"hints_used"`
	LastAttempt int64  `json:"last_attempt"` // unix timestamp
}

// ProgressStore manages progress persistence
type ProgressStore struct {
	filePath string
	progress []Progress
}

// NewProgressStore creates a new progress store
func NewProgressStore() *ProgressStore {
	homeDir, _ := os.UserHomeDir()
	storeDir := filepath.Join(homeDir, ".learn-go-tui")

	return &ProgressStore{
		filePath: filepath.Join(storeDir, "progress.json"),
		progress: []Progress{},
	}
}

// Load loads progress from file
func (ps *ProgressStore) Load() error {
	data, err := os.ReadFile(ps.filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil // No progress file yet
		}
		return fmt.Errorf("failed to read progress: %w", err)
	}

	if err := json.Unmarshal(data, &ps.progress); err != nil {
		return fmt.Errorf("failed to parse progress: %w", err)
	}

	return nil
}

// Save saves progress to file
func (ps *ProgressStore) Save() error {
	// Ensure directory exists
	dir := filepath.Dir(ps.filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	data, err := json.MarshalIndent(ps.progress, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal progress: %w", err)
	}

	if err := os.WriteFile(ps.filePath, data, 0644); err != nil {
		return fmt.Errorf("failed to write progress: %w", err)
	}

	return nil
}

// GetProgress returns progress for a specific exercise
func (ps *ProgressStore) GetProgress(chapterID, exerciseID string) *Progress {
	for i := range ps.progress {
		p := &ps.progress[i]
		if p.ChapterID == chapterID && p.ExerciseID == exerciseID {
			return p
		}
	}
	return nil
}

// SaveProgress saves or updates progress for an exercise
func (ps *ProgressStore) SaveProgress(chapterID, exerciseID string, completed bool, attempts, hintsUsed int) {
	// Find existing progress
	for i := range ps.progress {
		p := &ps.progress[i]
		if p.ChapterID == chapterID && p.ExerciseID == exerciseID {
			p.Completed = completed || p.Completed
			if attempts > 0 {
				p.Attempts = attempts
			}
			if hintsUsed > 0 {
				p.HintsUsed = hintsUsed
			}
			p.LastAttempt = time.Now().Unix()
			return
		}
	}

	// Add new progress
	ps.progress = append(ps.progress, Progress{
		ChapterID:   chapterID,
		ExerciseID:  exerciseID,
		Completed:   completed,
		Attempts:    attempts,
		HintsUsed:   hintsUsed,
		LastAttempt: time.Now().Unix(),
	})
}

// GetCompletedExercises returns all completed exercise IDs for a chapter
func (ps *ProgressStore) GetCompletedExercises(chapterID string) []string {
	var completed []string
	for _, p := range ps.progress {
		if p.ChapterID == chapterID && p.Completed {
			completed = append(completed, p.ExerciseID)
		}
	}
	return completed
}

// IsCompleted checks if an exercise is completed
func (ps *ProgressStore) IsCompleted(chapterID, exerciseID string) bool {
	p := ps.GetProgress(chapterID, exerciseID)
	return p != nil && p.Completed
}

File: rust/src/progress.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Progress {
    pub chapter_id: String,
    pub exercise_id: String,
    pub completed: bool,
    pub attempts: i32,
    pub hints_used: i32,
    pub last_attempt: i64,
}

pub struct ProgressStore {
    file_path: PathBuf,
    progress: Vec<Progress>,
}

impl ProgressStore {
    pub fn new() -> Self {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
        let store_dir = home.join(".learn-rust-tui");
        Self {
            file_path: store_dir.join("progress.json"),
            progress: Vec::new(),
        }
    }

    pub fn load(&mut self) -> Result<(), String> {
        match fs::read_to_string(&self.file_path) {
            Ok(data) => {
                self.progress =
                    serde_json::from_str(&data).map_err(|e| format!("failed to parse progress: {e}"))?;
                Ok(())
            }
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(e) => Err(format!("failed to read progress: {e}")),
        }
    }

    pub fn save(&self) -> Result<(), String> {
        let dir = self.file_path.parent().ok_or("invalid progress path")?;
        fs::create_dir_all(dir).map_err(|e| format!("failed to create directory: {e}"))?;

        let data = serde_json::to_string_pretty(&self.progress)
            .map_err(|e| format!("failed to marshal progress: {e}"))?;

        fs::write(&self.file_path, data).map_err(|e| format!("failed to write progress: {e}"))?;
        Ok(())
    }

    pub fn save_progress(
        &mut self,
        chapter_id: &str,
        exercise_id: &str,
        completed: bool,
        attempts: i32,
        hints_used: usize,
    ) {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs() as i64)
            .unwrap_or(0);

        for p in &mut self.progress {
            if p.chapter_id == chapter_id && p.exercise_id == exercise_id {
                p.completed = completed || p.completed;
                if attempts > 0 {
                    p.attempts = attempts;
                }
                if hints_used > 0 {
                    p.hints_used = hints_used as i32;
                }
                p.last_attempt = now;
                return;
            }
        }

        self.progress.push(Progress {
            chapter_id: chapter_id.to_string(),
            exercise_id: exercise_id.to_string(),
            completed,
            attempts,
            hints_used: hints_used as i32,
            last_attempt: now,
        });
    }

    pub fn get_completed_exercises(&self, chapter_id: &str) -> Vec<String> {
        self.progress
            .iter()
            .filter(|p| p.chapter_id == chapter_id && p.completed)
            .map(|p| p.exercise_id.clone())
            .collect()
    }

    pub fn entries(&self) -> &[Progress] {
        &self.progress
    }
}
